app/lib/features/chat_ng/widgets/chat_input/chat_editor.dart

```diff
@@ -0,0 +1,306 @@
+        .every((node) => node.delta?.toPlainText().isEmpty ?? true);
+    _isInputEmptyNotifier.value = state;
+    _debounceTimer?.cancel();
+    // delay operation to avoid excessive re-writes
+    _debounceTimer = Timer(const Duration(milliseconds: 300), () async {
+      // save composing draft
+      final text = textEditorState.intoMarkdown();
+      final htmlText = textEditorState.intoHtml();
+      await saveDraft(text, htmlText, widget.roomId, ref);
+      _log.info('compose draft saved for room: ${widget.roomId}');
+    });
+  }
+
+  // handler for expanding editor field height
+  void _updateHeight() {
+    final text = textEditorState.intoMarkdown();
+    final lineCount = '\n'.allMatches(text).length;
+
+    // Calculate new height based on line count
+    // Start with 5% and increase by 4% per line up to 20%
+    setState(() {
+      _cHeight = (0.05 + (lineCount - 1) * 0.04).clamp(0.05, 0.15);
+    });
+  }
+
+  // composer draft load state handler
+  Future<void> _loadDraft() async {
+    final draft =
+        await ref.read(chatComposerDraftProvider(widget.roomId).future);
+
+    if (draft != null) {
+      final inputNotifier = ref.read(chatInputProvider.notifier);
+      inputNotifier.unsetSelectedMessage();
+      draft.eventId().map((eventId) {
+        final draftType = draft.draftType();
+        final m = ref
+            .read(chatMessagesProvider(widget.roomId))
+            .firstWhere((x) => x.id == eventId);
+        if (draftType == 'edit') {
+          inputNotifier.setEditMessage(m);
+        } else if (draftType == 'reply') {
+          inputNotifier.setReplyToMessage(m);
+        }
+      });
+
+      final transaction = textEditorState.transaction;
+      final doc = ActerDocumentHelpers.parse(
+        draft.plainText(),
+        htmlContent: draft.htmlText(),
+      );
+      Node rootNode = doc.root;
+      transaction.document.insert([0], rootNode.children);
+      transaction.afterSelection =
+          Selection.single(path: rootNode.path, startOffset: 0);
+      textEditorState.apply(transaction);
+
+      _log.info('compose draft loaded for room: ${widget.roomId}');
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final isKeyboardVisible = ref.watch(keyboardVisibleProvider).valueOrNull;
+    final emojiPickerVisible = ref
+        .watch(chatInputProvider.select((value) => value.emojiPickerVisible));
+    final viewInsets = MediaQuery.viewInsetsOf(context).bottom;
+    return Column(
+      children: <Widget>[
+        renderEditorUI(emojiPickerVisible),
+        // Emoji Picker UI
+        if (emojiPickerVisible) ChatEmojiPicker(editorState: textEditorState),
+        // adjust bottom viewport so toolbar doesn't obscure field when visible
+        if (isKeyboardVisible != null && isKeyboardVisible)
+          SizedBox(height: viewInsets + 50),
+      ],
+    );
+  }
+
+  // chat editor UI
+  Widget renderEditorUI(bool emojiPickerVisible) {
+    return FrostEffect(
```

Comment:
what is this, btw? We have that a few times but I never really understood what it is for and why have it? There seems be zero difference in most cases if it just wasn't there... or maybe my :eyes: aren't good enough to catch the difference?

Reply:
This is UI wrapper to provide blurry/frost effect to inner child container, I remember this was added to reflect designs of Chat UI.